FileInstaller/FileInstaller.hpp

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <vector>
+
+#ifdef _DEBUG
+#define DEBUG_MSG(str) do { std::wcout << L"[" << __FUNCTION__ << "] " << str << std::endl; } while( false )
+#else
+#define DEBUG_MSG(str) do { } while ( false )
+#endif
+
+enum FileInstallerStatus : uint32_t {
+	FILEINSTALLER_SUCCESS = 0,
+	
+	FILEUTILS_COPY_FILE_COPY_FAILED,
+	FILEUTILS_DELETE_FILE_DELETE_FAILED,
+	FILEUTILS_CREATE_DIRECTORY_CREATE_FAILED,
+	FILEUTILS_DELETE_DIRECTORY_REMOVE_FAILED,
+};
+
+class FileInstallerException : public std::exception
+{
+public:
+	FileInstallerException(FileInstallerStatus status) : status(status) {};
+	FileInstallerStatus get_status() const { return status; }
+
+private:
+	FileInstallerStatus status;
+};
+
+/*
+FileInstaller - Allows installation of files in a target directory.
+
+Notice: FileInstaller class doesn't support paths longer than MAX_PATH.
+*/
+class FileInstaller final {
+public:
+	FileInstaller(std::vector<std::wstring> &file_paths, std::wstring const &installation_dir);
+	virtual ~FileInstaller();
+	void install();
+
+	FileInstaller(const FileInstaller&) = delete;
+	FileInstaller& operator=(FileInstaller const&) = delete;
+
+private:
+	void _copy_file(std::wstring const &file_path);
+	void _delete_file(std::wstring const &file_path);
+	void _copy_files();
+	void _create_installation_dir();
+	
+	// This method is working by best effort.
+	FileInstallerStatus _delete_installation_dir();
+
+	// This method is working by best effort.
+	FileInstallerStatus _delete_installed_files();
+
+	/*
+	The revert method cant be called standalone because it will revert only things performed during the install method.
+	This is working by best effort.
+	*/
+	void _revert_installation();
+
+private:
+	std::vector<std::wstring> &m_file_paths;
```

Comment:
Storing vectors and string as references is dangerous, consider what happens in the following ..

```c++

std::unique_ptr<FileInstaller> installer_ptr; 
// some code ..
{
   std::vector<..> file_paths;
   std::wstring installation_dir;
   installer_ptr = make_unique<FileInstaller>(file_paths, installation_dir); 
  }
installer_ptr->install();
```

Reply:
Your'e right.
Removed all references from class members.
All method still contains references because there is no risk.